### PR TITLE
feat(acme): add support for certificate ordering

### PR DESCRIPTION
**Pull request by lonelysadness**
Hello,
I've started work on ACME certificate ordering, and it is functioning in my tests using the Desec provider with the DNS-01 challenge.

While the implementation could likely use further optimization, I wanted to share it here as a starting point. Apologies if any parts of the code are rough. Feedback and suggestions are welcome!

### Contributor's Note
- [x] I have added / updated documentation in `/docs` for any user-facing features or additions.
- [x] I have added / updated acceptance tests in `/fwprovider/tests` for any new or updated resources / data sources.
- [x] I have ran `make example` to verify that the change works as expected.

### Proof of Work

```
=== RUN   TestAccResourceACMECertificate
--- PASS: TestAccResourceACMECertificate (11.06s)
```

```
=== RUN   TestFindMatchingCertificate
--- PASS: TestFindMatchingCertificate (0.00s)
    --- PASS: TestFindMatchingCertificate/match_certificate_by_SAN (0.00s)
    --- PASS: TestFindMatchingCertificate/match_certificate_by_CN_when_SANs_empty (0.00s)
    --- PASS: TestFindMatchingCertificate/prefer_ACME_cert_over_Proxmox-generated (0.00s)
    --- PASS: TestFindMatchingCertificate/prefer_cert_with_more_matching_domains (0.00s)
    --- PASS: TestFindMatchingCertificate/no_matching_domains_-_prefer_ACME_over_Proxmox (0.00s)
    --- PASS: TestFindMatchingCertificate/empty_certificates_list (0.00s)
    --- PASS: TestFindMatchingCertificate/fallback_to_first_cert_when_all_are_Proxmox-generated (0.00s)
    --- PASS: TestFindMatchingCertificate/match_with_multiple_SANs (0.00s)
    --- PASS: TestFindMatchingCertificate/match_CN_with_X.509_subject_format_(slash_delimiter) (0.00s)
    --- PASS: TestFindMatchingCertificate/match_CN_with_complex_X.509_subject_format (0.00s)

=== RUN   TestIsProxmoxGeneratedCertificate
--- PASS: TestIsProxmoxGeneratedCertificate (0.00s)
    --- PASS: TestIsProxmoxGeneratedCertificate/Proxmox_in_issuer (0.00s)
    --- PASS: TestIsProxmoxGeneratedCertificate/PVE_in_issuer (0.00s)
    --- PASS: TestIsProxmoxGeneratedCertificate/Let's_Encrypt_issuer (0.00s)
    --- PASS: TestIsProxmoxGeneratedCertificate/nil_issuer (0.00s)
    --- PASS: TestIsProxmoxGeneratedCertificate/empty_issuer (0.00s)
```

### Community Note

- Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates #2157

### CODE REVIEW

**Comment by lonelysadness** [comment]
I think I have a good functional starting point, I'll wait for @bpg  opinion or someone else's. Thank you.

PS: Sorry for the spam with Gemini, I didn't see the mess it made in the PR.

**Comment by bpg-dev** [comment]
> I think I have a good functional starting point, I'll wait for @bpg opinion or someone else's. Thank you.
> 
> PS: Sorry for the spam with Gemini, I didn't see the mess it made in the PR.

All good, thanks for contributing this @lonelysadness! ❤️

It will take some time to go over the PR tho...

File: fwprovider/nodes/resource_acme_certificate_test.go [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+	"github.com/bpg/terraform-provider-proxmox/utils"
+)
+
+// TestAccResourceACMECertificate tests the ACME certificate resource.
+// Note: This test requires a properly configured ACME environment:
+// - Set PROXMOX_VE_ACC_ACME_ACCOUNT_NAME environment variable
+// - Set PROXMOX_VE_ACC_ACME_DOMAIN environment variable
+// - Set PROXMOX_VE_ACC_ACME_DNS_PLUGIN (optional, for DNS-01 challenge)
+// The test will be skipped if these are not set.
+func TestAccResourceACMECertificate(t *testing.T) {
```
Comment (bpg-dev) [inline fwprovider/nodes/resource_acme_certificate_test.go:43]: Could you add more details here or to the PR description how to set up a test environment to run the tests, what acme server to use, how to set it up, etc.?

Just links to the docs and brief description, enough to anyone else to set that up from scratch.

File: fwprovider/nodes/resource_acme_certificate_test.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+	dnsPlugin := utils.GetAnyStringEnv("PROXMOX_VE_ACC_ACME_DNS_PLUGIN")
+
+	if acmeAccount == "" || acmeDomain == "" {
+		t.Skip("Skipping ACME certificate test - set PROXMOX_VE_ACC_ACME_ACCOUNT_NAME and PROXMOX_VE_ACC_ACME_DOMAIN")
+	}
+
+	nodeName := te.NodeName
+	if nodeName == "" {
+		nodeName = "pve"
+	}
```
Comment (bpg-dev) [inline fwprovider/nodes/resource_acme_certificate_test.go]: this is redundant, `te.NodeName` defaults to `pve`

File: fwprovider/nodes/resource_acme_certificate_test.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+	if dnsPlugin != "" {
+		domainsConfig += `
+		plugin = "` + dnsPlugin + `"`
+	}
+
+	domainsConfig += `
+	}]`
+
+	te.AddTemplateVars(map[string]interface{}{
+		"NodeName":      nodeName,
```
Comment (bpg-dev) [inline fwprovider/nodes/resource_acme_certificate_test.go]: it is already defined

File: fwprovider/nodes/resource_acme_certificate.go [hunk 4]
```diff
@@ -0,0 +1,882 @@
...
+				return fmt.Errorf("no certificates found yet")
+			}
+
+			certificates = certs
+
+			return nil
+		},
+		retry.Attempts(30),                    // Maximum 30 attempts
+		retry.Delay(1*time.Second),            // Start with 1 second delay
+		retry.DelayType(retry.BackOffDelay),   // Use exponential backoff
```
Comment (bpg-dev) [inline fwprovider/nodes/resource_acme_certificate.go]: exponential backoff with 30 attempts is a way too long IMO,
prob. just a fixed delay (5 sec? 10 sec?) backoff?

File: fwprovider/nodes/resource_acme_certificate.go [hunk 5]
```diff
@@ -0,0 +1,882 @@
...
+		retry.MaxJitter(500*time.Millisecond), // Add jitter to prevent thundering herd
```
Comment (bpg-dev) [inline fwprovider/nodes/resource_acme_certificate.go]: not sure if that will prevent anything in this scenario 😅

File: fwprovider/nodes/resource_acme_certificate.go [hunk 6]
```diff
@@ -0,0 +1,882 @@
...
+		}
+	}
+
+	// Order a new certificate if force is true or other changes are made
+	force := proxmoxtypes.CustomBool(plan.Force.ValueBool())
+	orderReq := &nodes.CertificateOrderRequestBody{
+		Force: &force,
+	}
+
+	taskID, err := nodeClient.OrderCertificate(ctx, orderReq)
```
Comment (bpg-dev) [inline fwprovider/nodes/resource_acme_certificate.go]: looks like we'll send the order call regardless if there were changes or not, or value of the force flag.

should we add a condition here like `if plan.Force.ValueBool() || !plan.Account.Equal(state.Account) || !plan.Domains.Equal(state.Domains) {....`

File: fwprovider/nodes/resource_acme_certificate.go [hunk 7]
```diff
@@ -0,0 +1,882 @@
...
+	issuer := *cert.Issuer
+
+	return strings.Contains(issuer, "Proxmox") || strings.Contains(issuer, "PVE")
+}
+
+// findMatchingCertificate finds the certificate that matches the domains in the model.
+// It prioritizes ACME certificates (issued by certificate authorities like Let's Encrypt)
+// over Proxmox-generated certificates. When multiple certificates match the configured domains,
+// it returns the one with the most matching domains.
+func (r *acmeCertificateResource) findMatchingCertificate(
```
Comment (bpg-dev) [inline fwprovider/nodes/resource_acme_certificate.go:702]: this logic seems pretty complex
i'd like to see unit test covering matching scenarios with some edge cases

**Comment by lonelysadness** [comment]
Thanks for all your feedback. I was busy, I'll take a look in the week !

**Comment by lonelysadness** [comment]
I'm currently very busy, I'll work on it again ASAP. Thanks.

**Comment by bpg-dev** [comment]
Ok, I'll look into updating & merging this PR

**Review by bpg — APPROVED** [review]
LGTM! 🚀

**Comment by bpg-dev** [comment]
@all-contributors please add @lonelysadness for code
